content/academy/puppeteer_playwright/executing_scripts.md

```diff
@@ -0,0 +1,143 @@
+
+```marked-tabs
+<marked-tab header="Playwright" lang="javascript">
+import { chromium } from 'playwright';
+
+const browser = await chromium.launch({ headless: false });
+const page = await browser.newPage();
+
+await page.goto('https://google.com/');
+
+await page.evaluate(() => {
+    document.body.style.background = 'green';
+});
+
+await page.waitForTimeout(10000)
+
+await browser.close();
+</marked-tab>
+<marked-tab header="Puppeteer" lang="javascript">
+import puppeteer from 'puppeteer';
+
+const browser = await puppeteer.launch({ headless: false });
+const page = await browser.newPage();
+
+await page.goto('https://google.com/');
+
+await page.evaluate(() => {
+    document.body.style.background = 'green';
+});
+
+await page.waitForTimeout(10000)
+
+await browser.close();
+</marked-tab>
+```
+
+Here's what we see in the automated browser when we run this code:
+
+![Google with the background color changed to green]({{@asset puppeteer_playwright/images/green-google.webp}})
+
+## [](#using-variables-in-page-evaluate) Using variables in `page.evaluate()`
+
+Within our code, we generate a `randomString` in the Node.js context:
+
+```JavaScript
+const randomString = Math.random().toString(36).slice(2);
+```
+
+Now, let's say we want to change the title of the document to be this random string. In order to use this `randomString` variable in the callback of our `page.evaluate()`, we'll have to pass in a second parameter containing the variable.
+
+> It's best practice to make this second parameter an object, as you are usually passing more than one value in.
+
+```marked-tabs
+<marked-tab header="Playwright" lang="javascript">
+import { chromium } from 'playwright';
+
+const browser = await chromium.launch({ headless: false });
+const page = await browser.newPage();
+
+await page.goto('https://google.com/');
+
+const randomString = Math.random().toString(36).slice(2);
+
+await page.evaluate(({ randomString }) => {
+    document.querySelector('title').textContent = randomString;
+}, { randomString });
+
+await page.waitForTimeout(10000)
+
+await browser.close();
+</marked-tab>
+<marked-tab header="Puppeteer" lang="javascript">
+import puppeteer from 'puppeteer';
+
+const browser = await puppeteer.launch({ headless: false });
+const page = await browser.newPage();
+
+await page.goto('https://google.com/');
+
+const randomString = Math.random().toString(36).slice(2);
+
+await page.evaluate(({ randomString }) => {
+    document.querySelector('title').textContent = randomString;
+}, { randomString });
+
+await page.waitForTimeout(10000)
+
+await browser.close();
+</marked-tab>
+```
+
+Now, when we run this code, we can see the title change on the page's tab:
+
+![Google with the background color changed to green]({{@asset puppeteer_playwright/images/title-changed.webp}})
+
+## [](#next) Next up
```

Comment:
This lesson is really short. How about adding a section about [`page.exposeFunction`](https://playwright.dev/docs/api/class-page#page-expose-function) to it?

I would first show a similar error as before. Trying to run a Node.js function from the browser context and failing. And then using exposeFunction to make it runnable.

Then I would talk about the limits of crossing the browser-node barrier. That only serializable objects can cross it and that you can't send full class instances back and forth.

Reply:
Thinking about it, I think that `executing_scripts` should not be a lesson, but a section and it should have multiple lessons about the various eval functions, as well as `evaluateOnNewDocument` vs `addInitScript`. `addScriptTag` (and some Puppeteer alternative) should be discussed too. It can be somehow intertwined with the "Collecting data" lesson.